File: bookscanner/matching.py
"""Fuzzy matching of OCR'ed title-page text against a structured title list."""
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass

from rapidfuzz import fuzz


def normalize_text(text: str) -> str:
    if not text:
        return ""
    text = unicodedata.normalize("NFKC", text)
    text = text.lower()
    text = re.sub(r"[^\w\s]", " ", text, flags=re.UNICODE)
    text = re.sub(r"\s+", " ", text).strip()
    return text


@dataclass
class MatchResult:
    row_index: int
    title: str
    score: float


def _specificity_factor(choice: str, floor_tokens: int = 3) -> float:
    """Down-weight catalog entries that are only 1-2 words long.

    token_set_ratio alone gives a near-perfect score to any short catalog
    title whose words are simply a subset of a longer, noisy OCR blob - e.g.
    a generic stub entry like "Eventyr" or "Andersen" inside a catalog of
    H.C. Andersen books would score 100 against almost every photo in that
    collection. Real, specific titles are essentially never just 1-2 words,
    so this only discounts those trivial stubs; it is independent of the
    query length, so it never penalises a genuinely correct short title (e.g.
    "Om Sindets Lidenskaber", 3 words) just because the photo also captured
    some ordinary surrounding publisher/edition text.
    """
    choice_len = len(choice.split())
    return min(1.0, choice_len / floor_tokens)


def find_best_matches(
    ocr_text: str, titles: list[str], limit: int = 3
) -> list[MatchResult]:
    """Match noisy OCR text against a list of catalog titles.

    Uses token_set_ratio so that a short catalog title matching a subset of
    words within a longer, noisy OCR blob still scores highly - this is what
    makes the matching tolerant of imperfect OCR and extra text on the page
    (publisher, edition, subtitle, etc). The raw score is then scaled by a
    specificity factor (see `_specificity_factor`) so that trivially short,
    generic catalog entries don't outrank the actual best-matching title.
    """
    query = normalize_text(ocr_text)
    if not query:
        return []

    choices = {i: normalize_text(t) for i, t in enumerate(titles)}
    choices = {i: t for i, t in choices.items() if t}
    if not choices:
        return []

    scored = []
    for idx, choice in choices.items():
        raw_score = fuzz.token_set_ratio(query, choice)
        score = raw_score * _specificity_factor(choice)
        scored.append((score, idx))

    scored.sort(key=lambda pair: pair[0], reverse=True)
    return [
        MatchResult(row_index=idx, title=titles[idx], score=score)
        for score, idx in scored[:limit]
    ]


def classify(score: float, high: float, low: float) -> str:
    if score >= high:
        return "Sandsynlig dublet"
    if score >= low:
        return "Muligvis dublet – tjek manuelt"
    return "Ingen match"

File: docs/js/matching.js
// Fuzzy matching of OCR'ed title-page text against a structured title list.
// Client-side port of bookscanner/matching.py - keep both in sync if either
// changes. See that file for the rationale behind each step.

(function (global) {
  "use strict";

  function normalizeText(text) {
    if (!text) return "";
    text = text.normalize("NFKC").toLowerCase();
    text = text.replace(/[^\p{L}\p{N}\s]/gu, " ");
    text = text.replace(/\s+/g, " ").trim();
    return text;
  }

  function tokensOf(text) {
    return text.length ? text.split(" ") : [];
  }

  // Longest common subsequence length via a space-optimized DP.
  function lcsLength(a, b) {
    if (a.length === 0 || b.length === 0) return 0;
    if (a.length < b.length) {
      const tmp = a;
      a = b;
      b = tmp;
    }
    let prev = new Array(b.length + 1).fill(0);
    let curr = new Array(b.length + 1).fill(0);
    for (let i = 1; i <= a.length; i++) {
      for (let j = 1; j <= b.length; j++) {
        if (a[i - 1] === b[j - 1]) {
          curr[j] = prev[j - 1] + 1;
        } else {
          curr[j] = Math.max(prev[j], curr[j - 1]);
        }
      }
      [prev, curr] = [curr, prev];
    }
    return prev[b.length];
  }

  // Indel-distance-based similarity ratio (0-100), equivalent in spirit to
  // rapidfuzz's fuzz.ratio(): 2 * LCS(a, b) / (len(a) + len(b)).
  function ratio(a, b) {
    const totalLen = a.length + b.length;
    if (totalLen === 0) return 100;
    const lcs = lcsLength(a, b);
    return (2 * lcs / totalLen) * 100;
  }

  // token_set_ratio: tolerant of extra/noisy words on either side by
  // comparing the shared vocabulary against each side's leftover words.
  function tokenSetRatio(query, choice) {
    const t1 = new Set(tokensOf(query));
    const t2 = new Set(tokensOf(choice));
    const intersection = [...t1].filter((w) => t2.has(w)).sort();
    const diff1 = [...t1].filter((w) => !t2.has(w)).sort();
    const diff2 = [...t2].filter((w) => !t1.has(w)).sort();

    const sortedSect = intersection.join(" ");
    const sorted1to2 = [sortedSect, diff1.join(" ")].filter(Boolean).join(" ").trim();
    const sorted2to1 = [sortedSect, diff2.join(" ")].filter(Boolean).join(" ").trim();

    return Math.max(
      ratio(sortedSect, sorted1to2),
      ratio(sortedSect, sorted2to1),
      ratio(sorted1to2, sorted2to1)
    );
  }

  // Down-weight catalog entries that are only 1-2 words long so a generic
  // stub entry (e.g. "Eventyr" or "Andersen" in an Andersen-only catalog)
  // can't outscore the actual full title just because its few words happen
  // to appear in the noisy OCR text. Independent of query length, so a
  // genuinely short but correct title (>=3 words) is never penalised just
  // because the photo also captured ordinary publisher/edition text.
  function specificityFactor(choice, floorTokens) {
    floorTokens = floorTokens || 3;
    const choiceLen = tokensOf(choice).length;
    return Math.min(1, choiceLen / floorTokens);
  }

  function findBestMatches(ocrText, titles, limit) {
    limit = limit || 3;
    const query = normalizeText(ocrText);
    if (!query) return [];

    const scored = [];
    for (let i = 0; i < titles.length; i++) {
      const choice = normalizeText(String(titles[i]));
      if (!choice) continue;
      const rawScore = tokenSetRatio(query, choice);
      const score = rawScore * specificityFactor(choice);
      scored.push({ rowIndex: i, title: titles[i], score: score });
    }
    scored.sort((a, b) => b.score - a.score);
    return scored.slice(0, limit);
  }

  function classify(score, high, low) {
    if (score >= high) return "Sandsynlig dublet";
    if (score >= low) return "Muligvis dublet – tjek manuelt";
    return "Ingen match";
  }

  global.bookscannerMatching = {
    normalizeText,
    tokenSetRatio,
    specificityFactor,
    findBestMatches,
    classify,
  };
})(typeof window !== "undefined" ? window : globalThis);
